### PR TITLE
planner: fix row count estimation for partially pushed down selections

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25486

Problem Summary:

Wrong row count estimations when selections can be detached from DataSource and are kept in tidb.

### What is changed and how it works?

What's Changed:

Compute the correct selectivity using the really pushed down filters.

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- Fix row count estimation for partially pushed down selections

### CODE REVIEW

File: planner/core/testdata/plan_suite_in.json [hunk 1]
```diff
@@ -703,5 +703,11 @@
       "select e from t where e = ''",
       "select e from t where e != ''"
     ]
+  },
+  {
+    "name": "TestSelectionPartialPushDown",
+    "cases": [
+      "select * from t where a > 1 and b > 1"
```
Comment (qw4990) [inline planner/core/testdata/plan_suite_in.json:710]: Please add one more test case for `Selection + IndexReader`

Comment (qw4990) [reply planner/core/testdata/plan_suite_in.json:710]: Rest LGTM

**Review by winoros — APPROVED** [review]

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -1259,6 +1259,14 @@ func (is *PhysicalIndexScan) addPushedDownSelection(copTask *copTask, p *DataSou
 		copTask.finishIndexPlan()
 		copTask.cst += copTask.count() * sessVars.CopCPUFactor
 		tableSel := PhysicalSelection{Conditions: tableConds}.Init(is.ctx, finalStats, is.blockOffset)
+		if len(copTask.rootTaskConds) != 0 {
+			selectivity, _, err := copTask.tblColHists.Selectivity(is.ctx, tableConds, nil)
+			if err != nil {
+				logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
+				selectivity = SelectionFactor
+			}
+			tableSel.stats = copTask.plan().statsInfo().Scale(selectivity)
```
Comment (time-and-fate) [inline planner/core/find_best_task.go:1268]: I think applying the selectivity of `tableConds` to the stats of `IndexScan`, which has considered `accessCond` and `ExpectedCnt`, is not accurate enough. But I also don't have any simple and better solution.

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -1842,6 +1850,14 @@ func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *pro
 	if len(ts.filterCondition) > 0 {
 		copTask.cst += copTask.count() * sessVars.CopCPUFactor
 		sel := PhysicalSelection{Conditions: ts.filterCondition}.Init(ts.ctx, stats, ts.blockOffset)
+		if len(copTask.rootTaskConds) != 0 {
+			selectivity, _, err := copTask.tblColHists.Selectivity(ts.ctx, ts.filterCondition, nil)
+			if err != nil {
+				logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
+				selectivity = SelectionFactor
+			}
+			sel.stats = ts.stats.Scale(selectivity)
```
Comment (time-and-fate) [inline planner/core/find_best_task.go:1859]: ditto.

File: planner/core/task.go [hunk 4]
```diff
@@ -1036,7 +1038,12 @@ func (t *copTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
 	}
 
 	if len(t.rootTaskConds) > 0 {
-		sel := PhysicalSelection{Conditions: t.rootTaskConds}.Init(ctx, newTask.p.statsInfo(), newTask.p.SelectBlockOffset())
+		selectivity, _, err := t.tblColHists.Selectivity(ctx, t.rootTaskConds, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
+			selectivity = SelectionFactor
+		}
+		sel := PhysicalSelection{Conditions: t.rootTaskConds}.Init(ctx, newTask.p.statsInfo().Scale(selectivity), newTask.p.SelectBlockOffset())
```
Comment (time-and-fate) [inline planner/core/task.go:1046]: I think we can directly use the row count from `DataSource.stats` here.

**Review by time-and-fate — COMMENTED** [review]
And it seems we're not calculating the cost for `Selection` in these places?

File: planner/core/task.go (hunk 4 above)
Comment (eurekaka) [reply planner/core/task.go:1046]: We don't have the `DataSource.stats` stored in `copTask` yet?

**Comment by eurekaka** [comment]
> And it seems we're not calculating the cost for `Selection` in these places?

The cost computation for selection is not very related to the changes of this PR, I prefer to do it in a separate PR.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**